util/stopper.go

```diff
@@ -17,39 +17,146 @@
+	return atomic.LoadInt32(&s.state)
+}
+
+// AddWorker worker to the stopper.
+func (s *Stopper) AddWorker() {
+	// If we're already draining, we might already be waiting on the stop
+	// WaitGroup, and it's a data race to add to it now. We hold the lock
+	// while waiting, so obtaining the lock here prevents that race.
+	// This means that possibly we only add to the WaitGroup after the
+	// stopper has shut down, but Closers added when the stopper is not are
+	// closed immediately, so the worker will stop anyways.
+	//
+	// This will likely never happen outside of tests, where we sometimes
+	// create and stop components in rapid succession.
+	s.mu.Lock()
+	s.stop.Add(1)
+	s.mu.Unlock()
+}
+
+// AddCloser adds an object to close after the stopper has been stopped.
+func (s *Stopper) AddCloser(c Closer) {
+	// Hold the lock to make sure we don't cut through a half-way Stop().
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	// If the state is not RUNNING, and we have the lock, then we should already
+	// done closing the stoppers in s.closers. The right thing to do is to just
+	// close it directly, and to never add it to the slice.
+	if s.getState() != RUNNING {
+		c.Close()
+		return
 	}
-	s.Add(count)
-	return s
+	// Otherwise, the closers are not yet being processed, so we can add it.
+	s.closers = append(s.closers, c)
 }
 
-// Add a new goroutine to the stopper.
-func (s *Stopper) Add(count int) {
-	s.wg.Add(count)
+// StartTask adds one to the count of tasks left to drain in the
+// system. Any worker which is a "first mover" when starting tasks
+// must call this method before starting work on a new task and must
+// subsequently invoke FinishTask() when the task is complete.
+// First movers include goroutines launched to do periodic work and
+// the kv/db.go gateway which accepts external client
+// requests.
+//
+// Returns true if the task can be launched or false to indicate the
+// system is currently draining and the task should be refused.
+func (s *Stopper) StartTask() bool {
+	if s.getState() == RUNNING {
+		s.drain.Add(1)
+		return true
+	}
+	return false
 }
 
-// Stop signals the waiting goroutine to stop and then waits
-// for it to confirm it has stopped (it does this by calling
-// SetStopped).
+// FinishTask removes one from the count of tasks left to drain in the
+// system. This function must be invoked for every successful call to
+// StartTask().
+func (s *Stopper) FinishTask() {
+	s.drain.Done()
+}
+
+// Stop signals all live workers to stop and then waits for each to
+// confirm it has stopped (workers do this by calling SetStopped()).
 func (s *Stopper) Stop() {
+	// Hold the lock all the time, so that AddWorker can't add to the
+	// stop WaitGroup while we Wait().
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	// If the state isn't RUNNING, Stop() has been invoked before.
+	// Since we are holding the lock, it must have completed.
+	if s.getState() != RUNNING {
+		return
+	}
+	s.setState(DRAINING)
+	s.drain.Wait()
```

Comment:
Ah, here is spot where you wait for draining to clear while holding the lock. I think it would be better for `Stopper.drain` and `Stopper.stop` to be integers and to use a condition variable to wit for them to reach the desired values.